ambari-server/src/main/java/org/apache/ambari/server/controller/AmbariManagementControllerImpl.java

```diff
@@ -751,6 +752,37 @@ public synchronized void createHostComponents(Set<ServiceComponentHostRequest> r
       throw new DuplicateResourceException(msg + names);
     }
 
+    //TODO move to separate method
+    for (String clusterName : hostComponentNames.keySet()) {
+      for (String serviceName : hostComponentNames.get(clusterName).keySet()) {
+        for (String componentName : hostComponentNames.get(clusterName).get(serviceName).keySet()) {
+          Set<String> hostnames = hostComponentNames.get(clusterName).get(serviceName).get(componentName);
+          if (hostnames != null && !hostnames.isEmpty()) {
+            ServiceComponent sc = clusters.getCluster(clusterName).getService(serviceName).getServiceComponent(componentName);
+            StackId stackId = sc.getDesiredStackId();
+            List<DependencyInfo> dependencyInfos = ambariMetaInfo.getComponentDependencies(stackId.getStackName(),
+                                                               stackId.getStackVersion(), serviceName, componentName);
+            for (DependencyInfo dependencyInfo : dependencyInfos) {
+              if ("host".equals(dependencyInfo.getScope()) && "exclusive".equals(dependencyInfo.getType())) {
+                Service depSevice = clusters.getCluster(clusterName).getService(dependencyInfo.getServiceName());
+                if (depSevice != null) {
+                  ServiceComponent dependentSC = depSevice.getServiceComponent(dependencyInfo.getComponentName());
+                  if (dependentSC != null) {
+                    for (String host : dependentSC.getServiceComponentHosts().keySet()) {
+                      if (hostnames.contains(host)) {
```

Comment:
I'd prefer using set operations instead of a loop: may be a bit faster, but more importantly, the error message can include all hosts with the problem, so it can be fixed in one go.